File: src/main/java/me/dags/copy/brush/cloud/Cloud.java
package me.dags.copy.brush.cloud;

import com.flowpowered.math.vector.Vector2f;
import com.flowpowered.math.vector.Vector3i;
import com.google.common.collect.ImmutableList;
import com.google.common.util.concurrent.FutureCallback;
import java.util.List;
import java.util.UUID;
import me.dags.copy.block.volume.BufferBuilder;
import me.dags.copy.block.volume.BufferView;
import me.dags.copy.brush.option.OptionHolder;
import org.spongepowered.api.block.BlockState;

/**
 * @author dags <dags@dags.me>
 */
public class Cloud {

    private final int seed;
    private final int octaves;
    private final double frequency;
    private final double max;
    private final int radius;
    private final int radius2;
    private final float featherDist2;
    private final float featherStart2;
    private final float opacity;
    private final int materialAir;
    private final int materialRange;
    private final float scaleFrequency;
    private final int heightRange;
    private final float center;
    private final float incline;
    private final Vector2f facing;

    private final List<BlockState> materials;

    public Cloud(OptionHolder brush, ImmutableList<BlockState> materials, Vector2f facing) {
        this.seed = brush.getOption(CloudBrush.SEED);

        this.frequency = brush.getOption(CloudBrush.FREQUENCY) * 0.1D;
        this.octaves = brush.getOption(CloudBrush.OCTAVES);
        this.max = ValueNoise.maxValue(octaves);

        this.radius = brush.getOption(CloudBrush.RADIUS);
        this.radius2 = radius * radius;

        float feather = brush.getOption(CloudBrush.FEATHER);
        this.featherDist2 = (radius * feather) * (radius * feather);
        this.featherStart2 = radius2 - featherDist2;

        this.opacity = brush.getOption(CloudBrush.OPACITY);

        float density = brush.getOption(CloudBrush.DENSITY);
        this.materialAir = Math.round(materials.size() * (1 - density));
        this.materialRange = materials.size() - 1 + materialAir;

        float scale = brush.getOption(CloudBrush.SCALE);
        this.scaleFrequency = (1 - scale) * 0.125F;

        int height = brush.getOption(CloudBrush.HEIGHT);
        this.heightRange = (2 * height) + 1;

        this.center = brush.getOption(CloudBrush.CENTER);
        this.incline = brush.getOption(CloudBrush.INCLINE);
        this.facing = facing;

        this.materials = materials;
    }

    public Runnable createTask(UUID owner, Vector3i pos, FutureCallback<BufferView> callback) {
        return () -> {
            try {
                BufferView volume = apply(pos, owner);
                callback.onSuccess(volume);
            } catch (Throwable t) {
                callback.onFailure(t);
            }
        };
    }

    public BufferView apply(Vector3i pos, UUID owner) {
        int diameter = 1 + radius * 2;
        int volume = diameter * heightRange * diameter;
        BufferBuilder buffer = new BufferBuilder(owner, Vector3i.ZERO, volume);

        for (int dz = 0; dz <= radius; dz++) {
            for (int dx = 0; dx <= radius; dx++) {
                int dist2 = (dx * dx) + (dz * dz);
                if (dist2 < radius2) {
                    apply(buffer, pos, dx, dz, dist2);
                    if (dx != 0) {
                        apply(buffer, pos, -dx, dz, dist2);
                    }
                    if (dz != 0) {
                        apply(buffer, pos, dx, -dz, dist2);
                    }
                    if (dx != 0 && dz != 0) {
                        apply(buffer, pos, -dx, -dz, dist2);
                    }
                }
            }
        }

        return buffer.getView();
    }

    private void apply(BufferBuilder buffer, Vector3i pos, int dx, int dz, float distance2) {
        int x = pos.getX() + dx;
        int y = pos.getY();
        int z = pos.getZ() + dz;

        // amount of vertical offset based on distance from center, incline, and facing x,z ratio
        int rx = Math.round(dx * incline * facing.getX());
        int rz = Math.round(dz * incline * facing.getY());

        // height based on perlin(x,z) modified by distance from center & getFeatherMod amount
        double hMod = modifier(distance2, radius2);
        float feather = getFeatherModifier(distance2);
        float elevation = getElevation(x, z, hMod * feather);

        int minY = -Math.max(1, Math.round(elevation * center));
        int maxY = Math.round(minY + elevation);
        for (int dy = minY; dy <= maxY; dy++) {
            BlockState material = getMaterial(x, dy, z, minY, maxY, feather);
            buffer.addAbsolute(material, x, y + dy + rx + rz, z);
        }
    }

    private float getElevation(double x, double z, double mod) {
        double noise = ValueNoise.getValue(x, 0, z, seed, frequency, octaves) / max;
        return (float) (heightRange * noise * mod);
    }

    private float getFeatherModifier(float dist2) {
        if (dist2 > featherStart2) {
            return 1F - (dist2 - featherStart2) / featherDist2;
        }
        return 1F;
    }

    private BlockState getMaterial(double x, double y, double z, double lower, double upper, double feather) {
        double mod = modifier(y, y < 0.5 ? lower : upper);
        double noise = ValueNoise.getValue(x, y, z, seed, scaleFrequency, octaves) / max;
        double value = (materialRange * noise * feather * mod) - materialAir;
        int index = Math.min(materialRange, Math.max(0, (int) Math.round(opacity * value)));
        return materials.get(index);
    }

    private static double modifier(double value, double range) {
        double modifier = value / range;
        return 1D - (modifier * modifier);
    }
}

File: docs/brush/cloud-preview/cloud.js
let shape = new Circle();
let frequency = 64;
let heightFrequency = 1 / frequency;
let octaves = 3;

let height = 16;
let heightRange = (2 * height) + 1;
let density = 0.5;
let opacity = 0.9;
let scale = 0.3;
let scaleFrequency = scale * 0.1;
let center = 0.5;
let xdirection = 0;
let rotation = 0.5;

let maxNoise = 1;
let range = 255;
let air = 0;
let bitrate = 15;

let xOff = 0;
let zOff = 0;

function draw(buffer) {
  shape.iterate(perform);
}

function perform(buffer, x, z, dist2) {
  let px = x + xOff;
  let pz = z + zOff;

  let hmod = mod(dist2, shape.radius2);
  let feather = getFeather(dist2, shape.featherRadius, shape.featherRange);
  let elevation = getElevation(px, pz, hmod * feather);

  let lower = -Math.max(1, Math.round(elevation * center));
  let upper = lower + elevation;

  let plan = getAlpha(px, 0, pz, lower, upper, feather);
  buffer.drawPlan(x, z, 255, 255, 255, plan);

  if (z === 0) {
    for (let dy = lower; dy <= upper; dy++) {
      let section = getAlpha(px, dy, pz, lower, upper, feather);
      buffer.drawSection(x, -dy, 255, 255, 255, section);
    }
  }
}

function getElevation(x, z, hmod) {
  let noise = getValue(x, 0, z, heightFrequency, octaves) / maxNoise;
  return noise * heightRange * hmod;
}

function getAlpha(x, y, z, lower, upper, feather) {
  let vmod = mod(y, y < 0.5 ? lower : upper);
  let noise = getValue(x, y, z, scaleFrequency, octaves) / maxNoise;
  let alpha = (range * noise * feather * vmod) - air;
  return clampAlpha(alpha, bitrate);
}

function getFeather(val, bound, range) {
  if (val > bound) {
    let d = val - bound;
    return 1 - (d / range);
  }
  return 1;
}

function clampAlpha(value, bitrate) {
    let max = 255.0;
    let perc = value / 255.0;
    let steps = Math.round(bitrate * perc);
    return Math.min(255, Math.max(0, opacity * steps * Math.round(255 / bitrate)));
}

function mod(val, range) {
  let mod = val / range;
  return 1 - (mod * mod);
}

function setPos(x, z) {
  xOff = x;
  zOff = z;
}

function registerOptions(register) {
  register(createSlider('frequency', 1, 100, 25, 100, (val) => {
    frequency = val;
    heightFrequency = frequency * 0.1;
  }));
  register(createSlider('octaves', 1, 6, 4, 1, (val) => {
    octaves = val;
    maxNoise = maxValue(octaves);
  }));
  register(createSlider('radius', 8, 70, 45, 1, (val) => shape.radius = val));
  register(createSlider('feather', 0, 100, 50, 100, (val) => shape.feather = val));
  register(createSlider('opacity', 0, 100, 100, 100, (val) => {
      opacity = val;
    }));
  register(createSlider('density', 0, 100, 90, 100, (val) => {
    density = val;
    air = 255 * (1 - density);
    range = 255 + air;
  }));
  register(createSlider('scale', 0, 100, 50, 100, (val) => {
    scale = val;
    scaleFrequency = (1 - scale) * 0.1;
  }));
  register(createSlider('height', 4, 48, 16, 1, (val) => {
    height = val;
    heightRange = (2 * height) + 1;
  }));
  register(createSlider('center', 0, 100, 25, 100, (val) => center = val));
}

function writeOutput(output) {
  output.value = `/set ${frequency};${octaves};${shape.radius};${shape.feather};${opacity};${density};${scale};${height};${center}`;
}
